state/upgrades.go

```diff
@@ -504,3 +504,92 @@ func UpgradeNoProxyDefaults(st *State) error {
 	}
 	return nil
 }
+
+// AddNonDetachableStorageMachineId sets the "machineid" field on
+// volume and filesystem docs that are inherently bound to that
+// machine.
+func AddNonDetachableStorageMachineId(st *State) error {
+	return runForAllModelStates(st, addNonDetachableStorageMachineId)
+}
+
+func addNonDetachableStorageMachineId(st *State) error {
+	var ops []txn.Op
+	volumes, err := st.volumes(
+		bson.D{{"machineid", bson.D{{"$exists", false}}}},
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, v := range volumes {
+		var pool string
+		if v.doc.Info != nil {
+			pool = v.doc.Info.Pool
+		} else if v.doc.Params != nil {
+			pool = v.doc.Params.Pool
+		}
+		detachable, err := isDetachableVolumePool(st, pool)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if detachable {
+			continue
+		}
+		attachments, err := st.VolumeAttachments(v.VolumeTag())
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if len(attachments) != 1 {
```

Comment:
maybe add a comment about this along the lines of multiple attachments meaning it must be detachable